jolt-core/src/r1cs/builder.rs

```diff
@@ -0,0 +1,1443 @@
+                Variable::Constant => panic!(),
+                Variable::Auxiliary(index) => assert_eq!(aux_index, index),
+            }
+            aux[aux_index] = (0..self.uniform_repeat)
+                .into_par_iter()
+                .map(|step_index| {
+                    let required_z_values: Vec<F> = aux_compute
+                        .flat_vars
+                        .iter()
+                        .map(|var| match var {
+                            Variable::Input(input_index) => {
+                                inputs[(*input_index).into()][step_index]
+                            }
+                            Variable::Auxiliary(inner_aux_index) => {
+                                debug_assert!(*inner_aux_index < aux_index);
+                                aux[*inner_aux_index][step_index]
+                            }
+                            _ => panic!(),
+                        })
+                        .collect();
+                    aux_compute.compute(&required_z_values)
+                })
+                .collect();
+        }
+
+        aux
+    }
+
+    /// Total number of rows used across all uniform constraints across all repeats. Repeat padded to 2, but repeat * num_constraints not, num_constraints not.
+    pub(super) fn uniform_repeat_constraint_rows(&self) -> usize {
+        self.uniform_repeat * self.uniform_builder.constraints.len()
+    }
+
+    pub(super) fn offset_eq_constraint_rows(&self) -> usize {
+        self.uniform_repeat
+    }
+
+    /// Total number of rows used across all repeated constraints. Not padded to nearest power of two.
+    pub(super) fn constraint_rows(&self) -> usize {
+        self.offset_eq_constraint_rows() + self.uniform_repeat_constraint_rows()
+    }
+
+    pub(super) fn uniform_repeat(&self) -> usize {
+        self.uniform_repeat
+    }
+
+    /// Materializes the uniform constraints into a single sparse (value != 0) A, B, C matrix represented in (row, col, value) format.
+    pub fn materialize_uniform(&self) -> UniformR1CS<F> {
+        self.uniform_builder.materialize()
+    }
+
+    pub fn materialize_offset_eq(&self) -> NonUniformR1CS<F> {
+        // (a - b) * condition == 0
+        // A: a - b
+        // B: condition
+        // C: 0
+
+        let mut eq = SparseEqualityItem::<F>::empty();
+        let mut condition = SparseEqualityItem::<F>::empty();
+
+        let constraint = &self.offset_equality_constraint;
+
+        let sorted_condition = constraint.condition.1.sorted_terms();
+        sorted_condition
+            .iter()
+            .filter(|term| matches!(term.0, Variable::Input(_) | Variable::Auxiliary(_)))
+            .for_each(|term| {
+                condition.offset_vars.push((
+                    self.uniform_builder.variable_to_column(term.0),
+                    constraint.condition.0,
+                    from_i64::<F>(term.1),
+                ))
+            });
+        if let Some(term) = sorted_condition.last() {
```

Comment:
I think this can be an unwrap? condition should never be empty

Reply:
Pivoted to LC::constant_term()